### PR TITLE
[swiftSyntax] Performance improvements for deserialising ByteTrees

### PR DESCRIPTION
**Pull request by ahoppen**
This PR contains several performance for deserialising syntax trees in binary encoding (see the separate commits):

- Make `SourceLength` a struct 
  - It should have been one all the while
- Make `RawSyntaxData` a direct enum
  - I do not see any reason why it shouldn't be
- Make `ByteTree(Object)Reader` a struct and pass a reference to it around using an `UnsafeMutabelPointer`
  - This improves the deserialisation performance by 13% because we are able to eliminate unnecessary retain and release calls
- Only reference `userInfo` through an `UnsafePointer` and don't pass it around
  - This improves deserialisation performance by another 11% because we are again able to eliminate unnecessary retain/release calls
- Use the more efficient array initialiser that was introduced in https://github.com/apple/swift/pull/17774

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: tools/SwiftSyntax/ByteTreeDeserialization.swift [hunk 1]
```diff
@@ -48,32 +48,34 @@ protocol ByteTreeObjectDecodable {
   ///   - numFields: The number of fields that are present in the serialized
   ///                object
   /// - Returns: The deserialized object
-  static func read(from reader: ByteTreeObjectReader, numFields: Int,
-                   userInfo: [ByteTreeUserInfoKey: Any]) -> Self
+  static func read(from reader: UnsafeMutablePointer<ByteTreeObjectReader>,
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply tools/SwiftSyntax/ByteTreeDeserialization.swift]: Because `inout ByteTreeObjectReader` still has a retain/release count associated with it.

**Review by nkcsgexi — APPROVED** [review]

**Comment by ahoppen** [comment]
Bots were unable to clone the projects. Let's try again.

@swift-ci Please smoke test and merge

**Comment by ahoppen** [comment]
Let's see if swift-ci is working again.

@swift-ci Please smoke test and merge

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Pull request merged**